### DISCUSSION TITLE
Welcome to JavaDoesUSB discussions!

### DISCUSSION THREAD

**Discussion by manuelbl** [manuelbl/JavaDoesUSB#2] [Announcements]
## 👋 Welcome!

We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions
  * Share ideas
  * Engage with other community members
  * Welcome others and are open-minded

  To get started, comment below and tell us what you are doing (or trying to do) with the *JavaDoesUSB* library.

**Comment by q2q-2q2** [comment]
for Fedora libudev installed via

```
dnf install libgudev libgudev-devel

**Reply by manuelbl** [reply]
Thanks for the feedback.

It `libgudev-devel` actually needed to run the Java code, or just for the C++ reference code?

**Reply by q2q-2q2** [reply]
Hi! Install binaries. Give permissions, as described in JavaDoesUSB manuals and run the java code. Everything will work. 
But after recent update Linux Kernel on Fedora to 6.2.x it cannot access device. Only enumerate. Have no idea why.

**Reply by manuelbl** [reply]
Which code are you running? One of the examples? What output do you get? And what do mean with "cannot access device"?

**Reply by q2q-2q2** [reply]
examples/bulk_transfer
I changed it to my PID and VID
and it fails with error "No such file or directory"
```  VID: 0x1a86
   PID: 0x7523
isOpen: false
isOpen: true
Exception in thread "main" net.codecrete.usb.linux.LinuxUSBException: Cannot claim USB interface. Нет такого файла или каталога (error code: 2)
	at net.codecrete.usb.linux.LinuxUSBException.throwException(LinuxUSBException.java:44)
	at net.codecrete.usb.linux.LinuxUSBException.throwLastError(LinuxUSBException.java:66)
	at net.codecrete.usb.linux.LinuxUSBDevice.claimInterface(LinuxUSBDevice.java:112)
	at net.codecrete.usb.examples.BulkTransfer.main(BulkTransfer.java:45)
```

this is my version of code
```
public class BulkTransfer {

    private static final int VID = 0x1a86;
    private static final int PID = 0x7523;
    private static final int INTERFACE_NO = 0;
    private static final int ENDPOINT_IN = 2;
    private static final int ENDPOINT_OUT = 2;

    public static void main(String[] args) {
        USBDevice device = Optional.ofNullable(USB.getDevice(new USBDeviceFilter(VID, PID))).orElseThrow();
        System.out.printf("   VID: 0x%04x%n", device.vendorId());
        System.out.printf("   PID: 0x%04x%n", device.productId());
        System.out.printf("isOpen: %b%n", device.isOpen());
        device.open();
        System.out.printf("isOpen: %b%n", device.isOpen());
        device.claimInterface(INTERFACE_NO);
        device.transferOut(ENDPOINT_OUT, new byte[] { 0x10 });```

**Reply by q2q-2q2** [reply]
examples/enumerate output
```Device:
  VID: 0x1a86
  PID: 0x7523
  Product name:  USB Serial
  Device class:    0xff (Vendor Specific Class)
  Device subclass: 0x00
  Device protocol: 0x00

  Interface 0
    Interface class:    0xff (Vendor Specific Class)
    Interface subclass: 0x01
    Interface protocol: 0x02

    Endpoint 2
        Direction: IN
        Transfer type: BULK
        Packet size: 32 bytes

    Endpoint 2
        Direction: OUT
        Transfer type: BULK
        Packet size: 32 bytes

    Endpoint 1
        Direction: IN
        Transfer type: INTERRUPT
        Packet size: 8 bytes
```

**Reply by q2q-2q2** [reply]
On windows 11 all works good

**Reply by manuelbl** [reply]
Your USB device is a USB-to-serial converter with a CH340G chip, right?

**Reply by q2q-2q2** [reply]
Not sure `libgudev-devel` need to be installed

**Reply by q2q-2q2** [reply]
Your USB device is a USB-to-serial converter with a CH340G chip, right?

Yes. Try to work with ModBus

**Reply by q2q-2q2** [reply]
After removing `libgudev-devel` on Fedora 37 it fails with exception
```
Exception in thread "main" net.codecrete.usb.USBException: Initial device enumeration has failed
	at net.codecrete.usb.common.USBDeviceRegistry.startDeviceMonitor(USBDeviceRegistry.java:136)
	at net.codecrete.usb.common.USBDeviceRegistry.start(USBDeviceRegistry.java:51)
	at net.codecrete.usb.USB.instance(USB.java:46)
	at net.codecrete.usb.USB.getDevice(USB.java:96)
	at net.codecrete.usb.examples.BulkTransfer.main(BulkTransfer.java:38)
Caused by: java.lang.ExceptionInInitializerError
	at net.codecrete.usb.linux.gen.udev.constants$0.<clinit>(constants$0.java:10)
	at net.codecrete.usb.linux.gen.udev.udev.udev_new$MH(udev.java:22)
	at net.codecrete.usb.linux.gen.udev.udev.udev_new(udev.java:25)
	at net.codecrete.usb.linux.LinuxUSBDeviceRegistry.monitorDevices(LinuxUSBDeviceRegistry.java:43)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.IllegalArgumentException: Cannot open library: libudev.so
	at java.base/java.lang.foreign.SymbolLookup.libraryLookup(SymbolLookup.java:235)
	at java.base/java.lang.foreign.SymbolLookup.libraryLookup(SymbolLookup.java:199)
	at net.codecrete.usb.linux.gen.udev.RuntimeHelper.<clinit>(RuntimeHelper.java:25)
	... 5 more
```

So, `libgudev-devel` is need to be installed too

**Reply by manuelbl** [reply]
I can reproduce the error with the USB to serial device. It occurs because a system driver is loaded for the device. The system driver can be unloaded and in a prototype implementation it already works.

More investigation is needed to come up with a suitable interface (should the driver be unloaded automatically or only with a separate operation, should the driver be reattached automatically etc.)

**Reply by manuelbl** [reply]
And regarding the Fedora packages: From what I can see, the minimal package to install would be `systemd-libs`.

**Reply by manuelbl** [reply]
The main branch now contains code to automatically unload the kernel driver. So your code for the CH340G should now be able to claim the interface. This works on Linux only. On Windows, the driver must be replaced with the WinUSB driver first. And on macos, I don't understand yet how the driver can be unloaded.

The main branch also contains a fix so that `libgudev-devel` is no longer needed on Linux. A default Linux installation should likely work.

**Reply by gospodbucar** [reply]
How to set little or big endian?

I also encountered a problem with interacting with device. It successfuly open and print - but i get an error trying to send or read endpoint (Exception in thread "main" net.codecrete.usb.windows.WindowsUsbException: transfer IN on endpoint 1 failed: An invalid parameter was passed to a service or function. (error code: -1073741811)) 

I was previously using java4usb library and i wanted to upgrade.

have any ideas?

**Reply by manuelbl** [reply]
Hi @gospodbucar 

What would you need a little ending / big ending setting for? Can you describe the use case?

The error you are getting could be a bug. I'd like to fix it but will need more information. Can you open a new issue with it. I would need:

- The code that leads to the error. If possible, a small but complete program. Otherwise sufficient code to reproduce it.
- If possible, the output of the [enumeration](https://github.com/manuelbl/JavaDoesUSB/tree/main/examples/enumerate) example program (the part related to the USB device that you want to communicate with is sufficient)

Am I correct assuming that you are using Java 21 and the latest version of this library?

**Comment by mattias-berglund** [comment]
Hi!

Thanks for the good job you've done creating this library, it suits my needs perfectly! However, I ran into a snag today when trying out my application on a Raspberry Pi:

```
Nov 13 12:46:01 raspberrypi launch.sh[2481]: Exception in thread "USB async IO" java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
Nov 13 12:46:01 raspberrypi launch.sh[2481]:         at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
Nov 13 12:46:01 raspberrypi launch.sh[2481]:         at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:124)
Nov 13 12:46:01 raspberrypi launch.sh[2481]:         at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:448)
Nov 13 12:46:01 raspberrypi launch.sh[2481]:         at java.base/java.util.Objects.checkIndex(Objects.java:441)
Nov 13 12:46:01 raspberrypi launch.sh[2481]:         at java.base/jdk.internal.foreign.LayoutPath.addScaledOffset(LayoutPath.java:243)
Nov 13 12:46:01 raspberrypi launch.sh[2481]:         at net.codecrete.usb.linux.LinuxAsyncTask.asyncCompletionTask(LinuxAsyncTask.java:101)
Nov 13 12:46:01 raspberrypi launch.sh[2481]:         at java.base/java.lang.Thread.run(Thread.java:1575)
```


It occurs if a try to handle two devices simultaneously, e.g. if two devices are available when the program starts. I'm running 1.1.0 on java23 (OpenJDK 64-Bit Server VM (build 23.0.1+11-snap, mixed mode, sharing)). Is there something I'm missing regarding thread safety in JavaDoesUSB or is it a bug?

**Reply by manuelbl** [reply]
Thanks for reporting it. I will look into it. The library is generally multi-threading safe. It looks like a bug.

**Comment by manuelbl** [comment]
I was able to reproduce the bug and fix it in version 1.1.1. It will soon be available on Maven Central.

**Reply by mattias-berglund** [reply]
Thanks for your swift response - now works like a charm!
